### PR TITLE
Add encoder functionality to Drive.kt

### LINKED ISSUE #4: Implement encoders on SRXes

### PR DESCRIPTION
**Pull request by ktorsh**
Resolves #4.

### CODE REVIEW

File: PowerUp-2018_test.iml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    <orderEntry type="library" name="Gradle: org.opencv:opencv-java:3.2.0" level="project" />
+    <orderEntry type="library" name="Gradle: edu.wpi.first.cscore:cscore-java:1.1.0" level="project" />
+    <orderEntry type="library" name="Gradle: openrio.mirror.third.ctre:CTRE-phoenix-java:5.1.2.1" level="project" />
+    <orderEntry type="library" name="Gradle: openrio.mirror.third.kauailabs:navx-java:3.0.342" level="project" />
+    <orderEntry type="library" name="Gradle: org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.4" level="project" />
+    <orderEntry type="library" name="Gradle: org.jetbrains.kotlin:kotlin-stdlib:1.2.10" level="project" />
+    <orderEntry type="library" name="Gradle: org.jetbrains:annotations:13.0" level="project" />
+  </component>
+  <component name="TestModuleProperties" production-module="PowerUp-2018_main" />
+</module>
```
Comment (plato2000) [inline PowerUp-2018_test.iml]: Can we untrack all of these high-churn project files?

File: src/main/java/org/usfirst/frc/team4099/lib/util/Rotation2D.kt [hunk 2]
```diff
@@ -1,121 +0,0 @@
-package org.usfirst.frc.team4099.lib.util
-
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/lib/util/Rotation2D.kt]: why was this file deleted?

File: src/main/java/org/usfirst/frc/team4099/lib/util/Utils.kt [hunk 3]
```diff
@@ -13,6 +13,9 @@ object Utils {
      * @param limit     limited magnitude
      * @return          the limited value
      */
+    fun convertSRXUnitstoPulses(unit: Double): Double{
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/lib/util/Utils.kt]: Style: should be `convertSRXUnitsToPulses` (capitalize To). Also, put a space before that curly brace.

On a different note, perhaps talk to @notjagan about that conversions interface and move this method there? Discuss using ft or wheel revs rather than just encoder revs.

File: src/main/java/org/usfirst/frc/team4099/robot/Constants.kt [hunk 4]
```diff
@@ -21,6 +35,18 @@ class Constants {
         val RIGHT_MASTER_ID = 4
         val RIGHT_SLAVE_1_ID = 5
         val RIGHT_SLAVE_2_ID = 0
+
+        val HIGH_GEAR_MAX_SETPOINT = 17.0 * 12.0 //17 fps
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/Constants.kt]: Is this how the max setpoints work in Phoenix? I thought they got rid of that 12V scaling factor

File: src/main/java/org/usfirst/frc/team4099/robot/Constants.kt [hunk 5]
```diff
@@ -21,6 +35,18 @@ class Constants {
...
+    }
+
+    object Wheels {
+        val DRIVE_WHEEL_DIAMETER_INCHES=0
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/Constants.kt]: please use spaces around equals signs

File: src/main/java/org/usfirst/frc/team4099/robot/Constants.kt [hunk 6]
```diff
@@ -21,6 +35,18 @@ class Constants {
...
+    }
+
+    object Velocity {
+        val highGearVelocityControlSlot=0
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/Constants.kt]: Same as above, also constants should be capitalized and underscore separated (there's a name for this but I can't remember what it is)

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 7]
```diff
@@ -1,14 +1,28 @@
...
 import com.kauailabs.navx.frc.AHRS
 import edu.wpi.first.wpilibj.SPI
 import edu.wpi.first.wpilibj.livewindow.LiveWindow
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
 import org.usfirst.frc.team4099.lib.drive.DriveSignal
 import org.usfirst.frc.team4099.robot.Constants
 import org.usfirst.frc.team4099.robot.loops.Loop
+import org.usfirst.frc.team4099.robot.subsystems.Drive.DriveControlState
+import com.ctre.CANTalon
+import sun.awt.windows.ThemeReader.getPosition
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: ?

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 8]
```diff
@@ -30,16 +49,125 @@ class Drive private constructor() : Subsystem {
...
         this.zeroSensors()
     }
 
+
+    @Synchronized
+    fun setOpenLoop(signal: DriveSignal) {
+        if (currentState !== DriveControlState.OPEN_LOOP) {
+            leftMasterSRX.set(ControlMode.PercentOutput, 0.0)
+            rightMasterSRX.set(ControlMode.PercentOutput, 0.0)
+            leftMasterSRX.configNominalOutputForward(0.0, 10)
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: I really don't like the timeout thing -- apparently, it's blocking and will hold the thread until it gets a response. It may be better to just give it 0 timeout and hope it went through. Is there ever a time where we change it from 0.0 to anything? If not, just give it zero timeout.

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 9]
```diff
@@ -30,16 +49,125 @@ class Drive private constructor() : Subsystem {
...
+        }
+        return false
+    }
+
+
+    fun getHighGear():Boolean {
+        return highGear
+    }
+
+    fun setHighGear(choice: Boolean){
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Can be replaced with Kotlin getters and setters (same for brake mode)

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 10]
```diff
@@ -30,16 +49,125 @@ class Drive private constructor() : Subsystem {
...
+    }
+
+
+
+
+    fun onStart(timestamp: Double){
+        synchronized (this) {
+            setOpenLoop(DriveSignal.NEUTRAL)
+            setBrakeMode(NeutralMode.Coast)
+            setVelocitySetpoint(0.0, 0.0) //could update in future
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Why are we putting it in velocity control mode to start?

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 11]
```diff
@@ -30,16 +49,125 @@ class Drive private constructor() : Subsystem {
...
+    fun setHighGear(choice: Boolean){
+        highGear=choice
+    }
+
+    fun getBrakeMode(): NeutralMode{
+        return brakeMode
+    }
+
+    fun setBrakeMode(type: NeutralMode){
+        if (brakeMode!=type) {
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: I've missed this in a couple of places, but please use spaces between your binary operators. IntelliJ can do reformat code which should automatically fix all of them.

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 12]
```diff
@@ -30,16 +49,125 @@ class Drive private constructor() : Subsystem {
...
+    }
+
+    override fun stop(){
+        synchronized(this){
+            setOpenLoop(DriveSignal.NEUTRAL)
+        }
+    }
+
+    fun resetEncoders(){
+
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Should implement (or put TODO)

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 13]
```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
         }
     }
 
+    @Synchronized
+    fun setVelocitySetpoint(left_inches_per_sec: Double, right_inches_per_sec: Double) {
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Consistent casing please -- camelCase is correct

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 14]
```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
...
+            }
+            else{
+                scale=1.0
+            }
+            leftMasterSRX.set(ControlMode.Velocity,inchesPerSecondToRpm(leftInchesPerSec*scale))
+            rightMasterSRX.set(ControlMode.Velocity,inchesPerSecondToRpm(rightInchesPerSec*scale))
+        }
+        else {
+            print("Incorrect Velocity Control Mode")
+            leftMasterSRX.set(ControlMode.Velocity,0.0)
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: spaces here too, please (after the comma)

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 15]
```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
...
+    }
+
+    @Synchronized
+    private fun updatePositionSetpoint(leftPositionInches: Double, rightPositionInches: Double){
+        if (usesTalonPositionControl(currentState)){
+            leftMasterSRX.set(ControlMode.MotionMagic,leftPositionInches)
+            leftMasterSRX.set(ControlMode.MotionMagic,leftPositionInches)
+        }
+        else {
+            print("Bad position control state")
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: `println`, otherwise things will get messy quick

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 16]
```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
...
+            leftMasterSRX.set(ControlMode.MotionMagic,0.0)
+            rightMasterSRX.set(ControlMode.MotionMagic,0.0)
+        }
+    }
+
+    private fun configureTalonsForVelocityControl() { //should further review cause im bad
+        if (!usesTalonVelocityControl(currentState)) {
+            // We entered a velocity control state.
+            leftMasterSRX.set(ControlMode.Velocity,0.0) //velocity  output value is in position change / 100ms
+            leftMasterSRX.configNominalOutputReverse(Constants.Velocity.driveHighGearNominalOutput, 10)
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Nominal output should (I think) always be zero, while peak output should actually be configurable. Nominal is minimum.

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 17]
```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
...
+            leftMasterSRX.selectProfileSlot(Constants.Velocity.highGearVelocityControlSlot, 0)
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Unsure if we should always be in the high gear profile (also theoretically, velocity control and position control shouldn't have different control slots so there should only be 2 populated at any time)

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 18]
```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
...
+            rightMasterSRX.configNominalOutputReverse(Constants.Velocity.driveHighGearNominalOutput, 10)
+            rightMasterSRX.selectProfileSlot(Constants.Velocity.highGearVelocityControlSlot, 0)
+            rightMasterSRX.configPeakOutputForward(Constants.Velocity.driveHighGearNominalOutput, 10)
+            setBrakeMode(NeutralMode.Brake)
+        }
+    }
+     private fun configureTalonsforPositionControl(){
+         if (!usesTalonPositionControl(currentState)) {
+             // We entered a position control state.
+             leftMasterSRX.set(ControlMode.MotionMagic,0.0);
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Is this a delta value, or does it try to set your position to zero from where it started? If the second one, this absolutely should not be zero unless you reset the encoders.

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 19]
```diff
@@ -72,6 +200,78 @@ class Drive private constructor() : Subsystem {
...
+             leftMasterSRX.configNominalOutputReverse(1.0, 10);
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: As before, nominal output should always be zero

File: src/main/java/org/usfirst/frc/team4099/lib/util/Rotation2D.kt (hunk 2 above)
Comment (azhong2002) [reply src/main/java/org/usfirst/frc/team4099/lib/util/Rotation2D.kt]: I think I mistakenly added it but it may have been the wrong one

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt (hunk 8 above)
Comment (ktorsh) [reply src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: So should I set all timeouts to 0?

File: src/main/java/org/usfirst/frc/team4099/lib/util/Utils.kt (hunk 3 above)
Comment (ktorsh) [reply src/main/java/org/usfirst/frc/team4099/lib/util/Utils.kt]: That method is never used, so Jagan told us to delete it.

**Comment by plato2000** [comment]
@ktorsh can you remove the other .iml files too? also the .ipr and the code style file in .idea

File: PowerUp-2018_test.iml [hunk 20]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (plato2000) [inline PowerUp-2018_test.iml]: still need to remove these high churn files

File: src/main/java/org/usfirst/frc/team4099/robot/Constants.kt [hunk 21]
```diff
@@ -22,6 +36,23 @@ class Constants {
...
+    object Wheels {
+        val DRIVE_WHEEL_DIAMETER_INCHES = 0
+    }
+
+    object Velocity {
+        val HIGH_GEAR_VELOCITY_CONTROL_SLOT = 0
+        val LOW_GEAR_VELOCITY_CONTROL_SLOT = 0
+        val DRIVE_HIGH_GEAR_NOMINAL_OUTPUT = 0.0 //percentage
+        val DRIVE_LOW_GEAR_NOMINAL_OUTPUT = 0.0 //percentage
+        val DRIVE_HIGH_GEAR_FORWARD_OUTPUT = 1.0 //percentage
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/Constants.kt]: Rename to `DRIVE_HIGH_GEAR_MAX_FORWARD_OUTPUT` or something similar to clarify that it's a max

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 22]
```diff
@@ -21,11 +25,17 @@ class Drive private constructor() : Subsystem {
     private val rightSlave1SRX: TalonSRX = TalonSRX(Constants.Drive.RIGHT_SLAVE_1_ID)
     private val rightSlave2SRX: TalonSRX = TalonSRX(Constants.Drive.RIGHT_SLAVE_2_ID)
     private val ahrs: AHRS
+
+    private var brakeMode: NeutralMode = NeutralMode.Brake//sets whether the break mode should be coast (no resistence) or by force
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: default to coast rather than brake

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 23]
```diff
@@ -34,16 +44,129 @@ class Drive private constructor() : Subsystem {
...
+    }
+
+    override fun stop() {
+        synchronized(this) {
+            setOpenLoop(DriveSignal.NEUTRAL)
+        }
+    }
+
+    @Synchronized
+    fun resetEncoders() {
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: what do these methods do?

File: src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt [hunk 24]
```diff
@@ -34,16 +44,129 @@ class Drive private constructor() : Subsystem {
...
+        leftMasterSRX.config_kI(1, Constants.Gains.LEFT_HIGH_KI, 0)
+        leftMasterSRX.config_kD(1, Constants.Gains.LEFT_HIGH_KD, 0)
+        leftMasterSRX.config_kF(1, Constants.Gains.LEFT_HIGH_KF, 0)
+
+
 
         rightSlave1SRX.set(ControlMode.Follower, Constants.Drive.RIGHT_MASTER_ID.toDouble())
         rightSlave2SRX.set(ControlMode.Follower, Constants.Drive.RIGHT_MASTER_ID.toDouble())
         rightMasterSRX.set(ControlMode.PercentOutput, 0.0)
+        rightMasterSRX.configSelectedFeedbackSensor(FeedbackDevice.QuadEncoder, 0, 0)
```
Comment (plato2000) [inline src/main/java/org/usfirst/frc/team4099/robot/subsystems/Drive.kt]: Can you put (in a comment) what some of these lines do? it's not immediately obvious what, for example, `setSensorPhase` does
